tools/harness-simulation/posix/sniffer_sim/sniffer.py

```diff
@@ -128,29 +134,67 @@ def _sniffer_main_loop(self):
             except socket.timeout:
                 continue
 
-            with self._mutex:
+            with self._nodeids_mutex:
                 denied_nodeids = self._denied_nodeids
 
             # Equivalent to RF enclosure
             if nodeid not in denied_nodeids:
                 self._pcap.append(data)
 
+    def TransferFile(self, request, context):
+        """ Transfer the capture file. """
+
+        # Validate and change the state
+        with self._state_mutex:
+            if self._state == CaptureState.NONE or (self._state & CaptureState.SYNC):
+                return sniffer_pb2.FilterNodesResponse(status=sniffer_pb2.OPERATION_ERROR)
+            self._state |= CaptureState.SYNC
+
+        with self._tshark_proc_mutex:
```

Comment:
`Stop` may be called before `TransferFile` has set stdout nonblocking, in such case _tshark_proc is already terminated. Not sure what would happen if `.stdout` is accessed on a terminated proc. 

I think we want to make sure that we start the `SYNC` mode and set stdout non-blocking in a atomic way. This would mean we should use one mutex for locking rather than two mutexes (i.e. _state_mutex and _tshark_proc)

Reply:
Actually `self._tshark_proc.stdout` is still accessable and `self._tshark_proc.stdout.fileno()` returns the file descriptor correctly after running `self._tshark_proc.terminate()`. The `tshark` process is a defunct process (aka. zombie process) at the moment before `self._tshark_proc.wait()` gets called. Only after `self._tshark_proc.wait()`, `self._tshark_proc` and its member variables will be deconstructed.

Reply:
According to the reviews above, I agree that `_tshark_proc_mutex` is not necessary and can be removed.

Reply:
And I think the lock for `self._state` is unnecessary as well, because `TransferFile` is always called right after `Start`, and `FilterNodes` is always called when the sniffer is working. The access of `self._state` in `Stop` will never collide with that in `TransferFile` or `FilterNodes`.